datapipelines/src/main/kotlin/io/customer/datapipelines/extensions/JsonExtensions.kt

```diff
@@ -43,3 +43,16 @@ private fun Any?.toSerializableJson(): JsonElement {
         else -> this.toJsonElement()
     }
 }
+
+/**
+ * Recursively replaces all `null` values in nested structure with [JsonNull],
+ * preserving structure and type. Useful for safely converting to `JsonElement`
+ * using `Any.toJsonElement()` extension.
+ */
+@Suppress("UNCHECKED_CAST")
+internal fun <T> T?.sanitizeNullsForJson(): T = when (this) {
+    null -> JsonNull as T
+    is Map<*, *> -> this.mapValues { (_, v) -> v.sanitizeNullsForJson() } as T
+    is List<*> -> this.map { it.sanitizeNullsForJson() } as T
+    else -> this
+}
```

Comment:
Another variation I had in mind was following, but it looked overly complicated so I stick to current implementation for simplicity

```kotlin
internal fun <T> T?.sanitizeNullsForJsons(): T = when (this) {
    null -> JsonNull as T

    is Map<*, *> -> this.takeIf {
        it.values.none { v -> v == null || v is Map<*, *> || v is List<*> }
    } ?: this.mapValues { (_, v) -> v.sanitizeNullsForJson() } as T

    is List<*> -> this.takeIf {
        it.none { e -> e == null || e is Map<*, *> || e is List<*> }
    } ?: this.map { it.sanitizeNullsForJson() } as T

    else -> this
}
```

Reply:
The third option is to add the fix directly to `analytics`. That's also a good option if we want to try it out.